agate/agate/i_message_retrieval.py

```diff
@@ -0,0 +1,27 @@
+from typing import Any, List
+
+
+class iMessageRetrieval:
```

Comment:
At the risk of bikeshedding, I wondered why you've called this `iMessageRetrieval`? My instinct is for something like `BaseMessageRetrieval` (or even `BaseMessageRetriever`), indicating this is a base class (i.e., should be subclassed).

I can't quite decide if you mean for this to be an abstract class or not. The design feels to me like you want it to be an abstract class but then you create instances in the tests. (I need to have a look at how `Mock()` works.)

Reply:
I suspect you instantiate `iMessageRetrieval` in the tests because the retriever has to be that kind of object to satisfy the type hint, but you could redefine the functions in `EmptyMessageRetrieval` too (or maybe define a `MockMessageRetrieval` if we expect more tests that need the mocking).

Reply:
Sorry, this is probably my instinct to try to implement interfaces/protocol classes in python whereas python just uses base classes for everything. (I'm doing a dependency injection talk in a few weeks, discussing the approach), I'll look further into python conventions and get back to you.

Reply:
If you want to impose that classes can be instantiated unless some functions are implemented, I think you can use the [Abstract Base Class module](https://docs.python.org/3/library/abc.html) in the standard library but I've never used it myself. I think what you've done is reasonably "Pythonic".

If this did become an abstract base class, then I think you'd have to use `EmptyMessageRetrieval` and override its functions the same way you're currently doing it.

Reply:
@warrickball  I found Protocols which seem to be the python method, I've changed the name and made it a protocol https://typing.readthedocs.io/en/latest/spec/protocol.html#protocols

Reply:
Interesting! This (and the comparison with abstract base classes) sounds like demo content to me. :wink: